src/compat/eme/close_session.ts

```diff
@@ -14,65 +14,99 @@
-    // to call another method on session to guess if
-    // session is closed or not.
-    observableTimer(1000).pipe(
-      mergeMap(() => {
-        const tryToUpdateSession$ = castToObservable(
-          session.update(new Uint8Array(1))
-        );
-        return tryToUpdateSession$.pipe(
-          // Update has resolved, so we can't know if session is closed
-          map(() => { throw new Error("Compat: Couldn't know if session is " +
-                                      "closed"); }),
-          catchError((err : unknown) => {
-            // The caught error can tell if session is closed
-            // (Chrome may throw this error)
-            if (err instanceof Error &&
-                err.message === "The session is already closed."
-            ) {
-              return observableOf(null);
-            }
-            // The `closed` promise may resolve, even if `close()` result has not
-            // (it may happen on Firefox). Wait for it and timeout after 1 second.
-            // TODO There is a subtle TypeScript issue there that made casting to
-            // a type-compatible type mandatory. If a more elegant solution can
-            // be found, it should be preffered.
-            const sessionIsClosed$ = castToObservable(session.closed as Promise<unknown>);
-            return observableRace(
-              sessionIsClosed$,
-              observableTimer(1000).pipe(
-                map(() => { throw new Error("Compat: Couldn't know if session is " +
-                                            "closed"); })
-              )
-            );
-          })
-        );
-      })
-    )
-  );
+export default function closeSession(
+  session: MediaKeySession | ICustomMediaKeySession
+): Promise<void> {
+  const timeoutCanceller = new TaskCanceller();
+
+  return PPromise.race([
+    session.close()
+      .then(() => { timeoutCanceller.cancel(); }),
+    // The `closed` promise may resolve, even if `close()` result has not
+    // (seen at some point on Firefox).
+    session.closed
+      .then(() => { timeoutCanceller.cancel(); }),
+    waitTimeoutAndCheck(),
+  ]);
+
+  /**
+   * If the session is not closed after 1000ms, try to communicate with the
+   * MediaKeySession and check if an error is returned.
+   * This is needed because on some browsers with poor EME implementation,
+   * knowing when a MediaKeySession is closed is actually a hard task.
+   *
+   * The returned Promise will never reject.
+   * @returns {Promise.<undefined>}
+   */
+  async function waitTimeoutAndCheck() : Promise<void> {
+    try {
+      await cancellableSleep(1000, timeoutCanceller.signal);
+      await tryUpdatingSession();
+    } catch (err) {
+      if (err instanceof CancellationError) { // cancelled == session closed
+        return;
+      }
+      const message = err instanceof Error ?
+        err.message :
+        "Unknown error made it impossible to close the session";
```

Comment:
Not really related to this PR, but maybe we can add an enum/object for error messages to make editing them easier ?